File: LPIS_1/src/main/java/com/example/lab/expression/Expression.java
package com.example.lab.expression;

import com.example.lab.FunctionContext;
import com.example.lab.SemanticChecker;
import com.example.lab.Type;
import com.example.lab.interpreter.InterpreterContext;

import java.util.List;
import java.util.Optional;

public interface Expression extends SemanticChecker {

    default Object runInterpreter(InterpreterContext context) {
        var type = getType(context);
        if (type == null) {
            runVoidInterpreter(context);
            return null;
        }
        return switch (type) {
            case ELEMENT -> runElementInterpreter(context);
            case LIST -> runListInterpreter(context);
        };
    }

    Type getType(FunctionContext context);

    default void runVoidInterpreter(InterpreterContext context) {
        throw new IllegalArgumentException("cls = " + getClass());
    }

    default int runElementInterpreter(InterpreterContext context) {
        throw new IllegalArgumentException("cls = " + getClass());
    }

    default List<Integer> runListInterpreter(InterpreterContext context) {
        throw new IllegalArgumentException("cls = " + getClass());
    }

    default Optional<Type> tryGetType() {
        return Optional.of(getType(null));
    }

}

File: LPIS_1/src/main/java/com/example/lab/expression/Expression.kt
package com.example.lab.expression

import com.example.lab.expression.provider.ExpressionProvider
import com.example.lab.interpreter.InterpreterContext

sealed interface Expression {

	fun getTypeOrNull(context: ExpressionProvider.Context): Type?
	fun getType(context: ExpressionProvider.Context): Type =
		getTypeOrNull(context) ?: throw NullPointerException("$this has not type")

	fun runInterpreter(context: InterpreterContext)

	enum class Type {
		ELEMENT, LIST, UNIT, NOTHING;
	}
}

interface ElementExpression : Expression {

	override fun getTypeOrNull(context: ExpressionProvider.Context): Expression.Type? = Expression.Type.ELEMENT

	fun runElement(context: InterpreterContext): Int

	override fun runInterpreter(context: InterpreterContext) {
		runElement(context)
	}
}

interface ListExpression : Expression {

	override fun getTypeOrNull(context: ExpressionProvider.Context): Expression.Type? = Expression.Type.LIST

	fun runList(context: InterpreterContext): List<Int>

	override fun runInterpreter(context: InterpreterContext) {
		runList(context)
	}
}

interface UnitExpression : Expression {

	override fun getTypeOrNull(context: ExpressionProvider.Context): Expression.Type? = Expression.Type.UNIT

	fun runUnit(context: InterpreterContext)

	override fun runInterpreter(context: InterpreterContext) {
		runUnit(context)
	}
}

interface NothingExpression : Expression {

	override fun getTypeOrNull(context: ExpressionProvider.Context): Expression.Type? = Expression.Type.NOTHING

	override fun runInterpreter(context: InterpreterContext) {
		TODO("Nothing")
	}
}
